### PR TITLE
Add ManSection enum, own title line struct and other minor updates

### PR DESCRIPTION
**Pull request by sondr3**
The title line, manpage section and compatibility settings come from clap-rs/clap#3174, where we decided on extracting out Roff-specific functionality to this crate. I also did a few minor changes, like removing the derive for `PartialEq` and `Eq` from the structs. The title line struct will make #7 obsolete, and is in my opinion a better solution as the order of the elements is important and is not just a list of strings.

cc @epage

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,16 @@
...
+
+# Remove Cargo.lock from gitignore if creating an executable, leave it for libraries
+# More information here https://doc.rust-lang.org/cargo/guide/cargo-toml-vs-cargo-lock.html
 Cargo.lock
+
+# These are backup files generated by rustfmt
+**/*.rs.bk
+
+# MSVC Windows builds of rustc generate these, which store debugging information
+*.pdb
```
Comment (epage) [inline .gitignore:16]: How come you are changing the `.gitignore`?

File: src/lib.rs [hunk 2]
```diff
@@ -1,28 +1,123 @@
...
+    /// Games
+    Games,
+    /// Miscellaneous (including macro packages and conventions), e.g. man(7), groff(7)
+    Miscellaneous,
+    /// System administration commands (usually only for root)
+    SystemAdministrationCommands,
+    /// Kernel routines [Non standard]
+    KernelRoutines,
+}
+
```
Comment (epage) [inline src/lib.rs:65]: Unless I'm missing something, this seems like man policy when this crate is focused on the underlying markup.

File: src/lib.rs [hunk 3]
```diff
@@ -1,28 +1,123 @@
...
             content: Vec::new(),
         }
     }
 
+    /// Date of the last nontrivial change to the manpage. Should be formatted
+    /// in `YYYY-MM-DD`.
+    pub fn date(mut self, date: impl Into<String>) -> Self {
+        self.title.date = Some(date.into());
+        self
+    }
```
Comment (epage) [inline src/lib.rs:100]: Should we accept `year, month, day` arguments so we can enforce formatting or eventually add a date/time lib dependency (haven't looked to see where the community is at with time vs chrono)

File: src/lib.rs [hunk 4]
```diff
@@ -1,28 +1,123 @@
...
+            ManSection::FileFormats => 5,
+            ManSection::Games => 6,
+            ManSection::Miscellaneous => 7,
+            ManSection::SystemAdministrationCommands => 8,
+            ManSection::KernelRoutines => 9,
+        }
+    }
+}
+
+pub struct Roff {
```
Comment (epage) [inline src/lib.rs:82]: If `Roff` is the entry point for this crate. please keep it first in the file so its easier to jump in and see what this crate does

File: src/lib.rs [hunk 5]
```diff
@@ -1,28 +1,123 @@
...
-pub struct Roff {
+/// Title line for a manpage.
+pub struct Title {
     title: String,
-    section: i8,
+    section: ManSection,
+    date: Option<String>,
+    source: Option<String>,
+    manual: Option<String>,
+}
```
Comment (epage) [inline src/lib.rs:10]: How much of this is roff vs man policy?

File: src/lib.rs [hunk 6]
```diff
@@ -33,13 +128,18 @@ impl Troffable for Roff {
...
+        // Compatibility settings:
+        //
+        // Set sentence_space_size to 0 to prevent extra space between sentences separated
+        // by a newline the alternative is to add \& at the end of the line
+        writeln!(&mut res, ".ss \\n[.ss] 0").unwrap();
+        // Disable hyphenation
+        writeln!(&mut res, ".nh").unwrap();
+        // Disable justification (adjust text to the left margin only)
+        writeln!(&mut res, ".ad l").unwrap();
+
```
Comment (epage) [inline src/lib.rs:142]: Should we be hard coding this or leaving them up for the caller to configure?

File: src/lib.rs [hunk 7]
```diff
@@ -1,28 +1,123 @@
 use std::fmt::Write;
 
-#[derive(PartialEq, Eq)]
-pub struct Roff {
+/// Title line for a manpage.
+pub struct Title {
```
Comment (epage) [inline src/lib.rs:4]: We should at least implement `Debug` and `Clone` on everything

File: src/lib.rs [hunk 8]
```diff
@@ -48,7 +148,6 @@ impl Troffable for Roff {
     }
 }
 
-#[derive(PartialEq, Eq)]
```
Comment (epage) [inline src/lib.rs:51]: How come you've been dropping `PartialEq, Eq` from these items?

File: src/lib.rs [hunk 9]
```diff
@@ -101,10 +200,14 @@ pub fn italic(input: &str) -> String {
...
 
-pub fn list<C1: Troffable, C2: Troffable>(header: &'_ [C1], content: &'_ [C2]) -> String {
+pub fn list<C1: Troffable, C2: Troffable>(header: &[C1], content: &[C2]) -> String {
     format!(".TP\n{}\n{}", header.render(), content.render())
 }
 
-fn escape(input: &str) -> String {
+pub fn escape(input: &str) -> String {
     input.replace("-", r"\-")
 }
```
Comment (epage) [inline src/lib.rs:209]: Why is `escape` public?  It seems like the API should be enforcing whats escaped or not

File: src/lib.rs (hunk 8 above)
Comment (sondr3) [reply src/lib.rs:51]: I don't really know why it's derived, I can't imagine it was used for something besides testing (the first full commit has them derived but not used) because comparing manpages for equality sounds very niche.

File: src/lib.rs (hunk 9 above)
Comment (sondr3) [reply src/lib.rs:209]: This was an oversight on my part, I forgot that everything is escaped when rendered, I'll fix this.

File: src/lib.rs (hunk 7 above)
Comment (sondr3) [reply src/lib.rs:4]: Good point!

File: src/lib.rs (hunk 6 above)
Comment (sondr3) [reply src/lib.rs:142]: Maybe, I really just took them from the manpage generator that Asciidoctor uses (which is what `git` uses for their manpages)

File: .gitignore (hunk 1 above)
Comment (sondr3) [reply .gitignore:16]: Force of habit, it carried over from my previous work on `roff-rs`, I can revert this if you want to.

File: src/lib.rs (hunk 3 above)
Comment (sondr3) [reply src/lib.rs:100]: That's probably a better way of doing it, good point. Not sure I'd like to pull in a full library for date changes here, we might do it in `clap_man` instead to fill these in instead.

File: src/lib.rs (hunk 4 above)
Comment (sondr3) [reply src/lib.rs:82]: Will fix!

**Comment by sondr3** [comment]
Your two comments about whether things should go here or in a more manpage specific crate was one of the reasons I ended up trying to do everything in `clap_man`, the inclusion of a title line means we're already targeting manpages and not "pure" `roff`. It is a _massive_ type-setting system ([here](https://man.archlinux.org/man/groff.7.en) is short reference of GNU variant of the format, you can see we're only really touching a tiny, tiny part of it), and has historically been used for lots of different things from books, articles and more.

Nowadays however I thing nobody uses it outside of manpages anymore, and while the `man` crate contains useful tools to build manpages for CLI tools this crate already targets a subset of `roff` specific to manpages, at least in my opinion. :sweat_smile: `roffman` that you mentioned in the PR in `clap` has [sections](https://github.com/vv9k/roffman/blob/master/src/lib.rs#L323) in it (and a lot more, I'm just not a big fan of its API).

Sorry, this is a lot of words for saying that in my head `roff-rs` is already a manpage specific crate :slightly_smiling_face:

File: .gitignore (hunk 1 above)
Comment (epage) [reply .gitignore:16]: Please do so.  imo person/machine specific ignores should be set in the personal ignores rather than checked into the shared ignores.

File: src/lib.rs (hunk 6 above)
Comment (epage) [reply src/lib.rs:142]: Eh, good enough.  We can always add it later (which would be a breaking change, but we're still early in roffs development)

File: src/lib.rs (hunk 8 above)
Comment (epage) [reply src/lib.rs:51]: I lean towards keeping them (slightly less important than `Clone` and `Debug`) though we should make sure we call this out as a breaking change.

File: src/lib.rs (hunk 2 above)
Comment (epage) [reply src/lib.rs:65]: > Your two comments about whether things should go here or in a more manpage specific crate was one of the reasons I ended up trying to do everything in `clap_man`, the inclusion of a title line means we're already targeting manpages and not "pure" `roff`. It is a _massive_ type-setting system ([here](https://man.archlinux.org/man/groff.7.en) is short reference of GNU variant of the format, you can see we're only really touching a tiny, tiny part of it), and has historically been used for lots of different things from books, articles and more.
> 
> Nowadays however I thing nobody uses it outside of manpages anymore, and while the `man` crate contains useful tools to build manpages for CLI tools this crate already targets a subset of `roff` specific to manpages, at least in my opinion. sweat_smile `roffman` that you mentioned in the PR in `clap` has [sections](https://github.com/vv9k/roffman/blob/master/src/lib.rs#L323) in it, though the name implies that it is specific for manpages, but not a huge fan of its API.
> 
> Sorry, this is a lot of words for saying that in my head `roff-rs` is already a manpage specific crate slightly_smiling_face It's again one of the reasons I've attempted a few different approaches to this, I haven't found a good solution for it yet besides creating my own crate for this.

My thought is to prefer to keep this crate man agnostic unless its getting in the way of getting `clap_man` done.

It looks like we could have the Section enum and the title formatting external to this crate and lose nothing *but* I'm not as knee deep in both crates as you are, so I defer to your judgement.

Comment (sondr3) [reply src/lib.rs:65]: Due to the current architecture this crate cannot be used to create anything besides manpages, so it is in a sense already specific to manpages and not agnostic. I'm starting to lean more towards having a more intermediate crate specifically for creating manpages, I have a prototype laying around that I could revive. Thoughts?

File: src/lib.rs [hunk 10]
```diff
@@ -109,7 +109,7 @@ impl Troffable for Section {
     fn render(&self) -> String {
         let mut res = String::new();
 
-        writeln!(&mut res, ".SH {}", self.title.to_uppercase()).unwrap();
+        writeln!(&mut res, ".SH \"{}\"", self.title.to_uppercase()).unwrap();
```
Comment (larswirzenius) [inline src/lib.rs:160]: This will fail if the title contains quote characters. I'm not sure how to fix that.

File: src/lib.rs [hunk 11]
```diff
@@ -130,6 +130,16 @@ impl Troffable for Roff {
 
         writeln!(&mut res, "{}", self.title.render()).unwrap();
 
+        // Compatibility settings:
```
Comment (larswirzenius) [inline src/lib.rs:133]: Why is this needed? What requires turning of justification and hyphenation?

I'm looking at this from a manual page generation point of view, and non-justified manual pages are unusual, and thus jarring to read.

File: src/lib.rs (hunk 10 above)
Comment (epage) [reply src/lib.rs:160]: Oddly, I'm not seeing any quote escaping in asciidoctor, only left/right quotes.  I wonder if that is an artifact of them only using left/right quotes

https://github.com/asciidoctor/asciidoctor/blob/main/lib/asciidoctor/converter/manpage.rb#L716

File: src/lib.rs (hunk 11 above)
Comment (epage) [reply src/lib.rs:133]: On the topic of compat settings, asciidoctor does https://github.com/asciidoctor/asciidoctor/blob/main/lib/asciidoctor/converter/manpage.rb#L61 (with links out for information)

File: src/lib.rs (hunk 10 above)
Comment (epage) [reply src/lib.rs:160]: For some more information on escaping, check out https://www.ibm.com/docs/en/aix/7.2?topic=t-troff-command (search for mm Strings)

Comment (larswirzenius) [reply src/lib.rs:160]: I've been reading the [groff(7)](https://manpages.debian.org/bullseye/groff/groff.7.en.html), and I can't see any way to escape quotes in control lines or macro arguments.

**Comment by larswirzenius** [comment]
I had a closer look at the manual page generation and the roff crate. I've pushed two branches that I think would be an improvement.

https://github.com/larswirzenius/roff-rs/tree/liw/roff-only changes roff to have an abstraction level of ROFF only, dropping manual page stuff. The [commit message](https://github.com/larswirzenius/roff-rs/commit/fefa52f169b07398f242f1fd28cec973c34d1097) has some more detailed explanations. The branch isn't ready to be merged, and needs some cleanup first.

I then ported @sondr3's `maual` page generation to my roff version (https://github.com/larswirzenius/clap/tree/liw/generate-man). This, too, probably needs cleanup.

What do you think? Is my approach of any interest?

**Comment by epage** [comment]
> What do you think? Is my approach of any interest?

I really like this!  By focusing on the roff structure, I feel like it makes clearer the intent of what is going on while giving the benefit of escaping and typing (compared to doing everything with raw strings).

File: src/lib.rs (hunk 11 above)
Comment (larswirzenius) [reply src/lib.rs:133]: The point about plain apostrophes is a good one, and I should add that to my roff crate version. However, I can't see any explanation of why hyphenation and justification should be turned off. The sentence space size I can live with: it removes the need to make sure sentences end with a newline after the terminating punctuation.

**Comment by sondr3** [comment]
I too like your approach and from the comment over the PR on clap you have far more experience working with manpages and the roff-format, so you can be the judge for this crate. Do you want to create a PR against my fork of clap so we can integrate your changes once you feel it's ready? :smile:

**Comment by larswirzenius** [comment]
I will do that later today or tomorrow, as soon as I've dug myself out from this horde of elves I'm buried under.

**Comment by sondr3** [comment]
Superseded by #15.
